assembly/buffer/index.ts

```diff
@@ -16,19 +26,89 @@ export class Buffer extends Uint8Array {
+    return assembleBuffer(__alloc(size, idof<ArrayBuffer>()), 0, size);
+  }
+
+  public static fromArrayBuffer(buffer: ArrayBuffer, byteOffset: i32 = 0, length: i32 = -1): Buffer {
+    length = select(buffer.byteLength, length, length < 0);
+    if (i32(byteOffset < 0) | i32(byteOffset > buffer.byteLength - length)) throw new RangeError(E_INDEXOUTOFRANGE);
+    if (length == 0) return new Buffer(0);
+
+    return assembleBuffer(changetype<usize>(buffer), <usize>byteOffset, <u32>length);
+  }
+
+  public static fromString(value: string, encoding: string = "utf8"): Buffer {
+    let buffer: ArrayBuffer;
+    if (encoding == "utf8" || encoding == "utf-8") {
+      buffer = String.UTF8.encode(value);
+    } else if (encoding == "utf16le") {
+      buffer = String.UTF16.encode(value);
+    } else if (encoding == "hex") {
+      buffer = Buffer.HEX.encode(value);
+    } else if (encoding == "ascii") {
+      buffer = Buffer.ASCII.encode(value);
+    } else {
+      throw new TypeError("Invalid string encoding.");
+    }
+
+    // assemble the buffer
+    return assembleBuffer(changetype<usize>(buffer), 0, buffer.byteLength);
+  }
+
+  public static fromArray<T extends ArrayBufferView>(value: T, offset: i32 = 0, length: i32 = -1): Buffer {
+    length = select(value.length, length, length < 0);
+    if (i32(offset < 0) | i32(offset > value.length)) throw new RangeError(E_INDEXOUTOFRANGE);
+    if (length > value.length - offset) throw new RangeError(E_INVALIDLENGTH);
+    if (length == 0) return new Buffer(0);
+    let arrayBuffer = __alloc(length, idof<ArrayBuffer>());
+    if (value instanceof Array<string>) {
+      for (let i = 0; i < length; i++) {
+        let index = i + offset;
+        let byteValue = parseFloat(unchecked(value[index]));
+        store<u8>(arrayBuffer + <usize>i, <u8>(isFinite(byteValue) ? <u8>byteValue : 0));
+      }
+    } else {
+      for (let i = 0; i < length; i++) {
+        let index = i + offset;
+        let element = isDefined(unchecked(value[index]))
+          ? unchecked(value[index])
+          : value[index];
+
+        if (isFloat(element)) {
+          store<u8>(arrayBuffer + <usize>i, <u8>(isFinite(element) ? <u8>element : 0));
+        } else {
+          store<u8>(arrayBuffer + <usize>i, <u8>element);
+        }
+      }
+    }
+
+    return assembleBuffer(arrayBuffer, 0, length);
+  }
 
-    // set the properties
-    store<usize>(result, __retain(buffer), offsetof<Buffer>("buffer"));
-    store<usize>(result, buffer, offsetof<Buffer>("dataStart"));
-    store<i32>(result, size, offsetof<Buffer>("byteLength"));
+  public static fromBuffer(source: Buffer): Buffer {
+    let length = source.byteLength;
+    let data = __alloc(length, idof<ArrayBuffer>()); // retains
+    memory.copy(data, source.dataStart, length);
+    return assembleBuffer(data, 0, length);
+  }
 
-    // return and retain
-    return changetype<Buffer>(result);
+  // @ts-ignore: Buffer returns on all valid branches
```

Comment:
We can suppress this error by returning `changetype<Buffer>(null)` at the bottom here after the `ERROR`. What is the right way to do this?